bukuserver/translations/__init__.py

```diff
@@ -0,0 +1,78 @@
+
+def replace_obsolete(text):
+    '''Removes *blank* obsolete entries, and restores re-added *blank* values from old obsolete ones'''
+    text = re.sub(OLD_BLANK, '', text)
+    for obsolete in re.finditer(OLD, text):
+        _ctxt, _id, _str = obsolete.groups()
+        ctxt_re = ('' if _ctxt is None else f'msgctxt "{re.escape(_ctxt)}"{_EOL}')
+        if m := re.search(f'^{ctxt_re}msgid "{re.escape(_id)}"{_EOL}msgstr "()"{_EOL}', text, re.MULTILINE):
+            text = (text[:m.start(1)] + _str + text[m.end(1):]).replace(obsolete.group(0), '', 1)
+    return text
+
+def translations_generate():
+    '''Generates and patches the messages.pot template file'''
+    pybabel().run(['', 'extract', '--no-wrap', f'--mapping-file={MAPPING}',
+                   '--keywords=_ _l _p:1c,2 _lp:1c,2 lazy_gettext', f'--output-file={TEMPLATE}', BUKUSERVER])
+    print(f'patching PO template file at {TEMPLATE}')
+    with open(TEMPLATE, encoding='utf-8') as fin:
+        text = fin.read()
+    for k, v in STRINGS.items():
+        text = re.sub(k, v, text, count=1, flags=re.MULTILINE)
+    with open(CUSTOM, encoding='utf-8') as fin:
+        with open(TEMPLATE, 'w', encoding='utf-8') as fout:
+            fout.write(text + fin.read())
+
+def translations_update(new_locales=[], generate=True, domain=DOMAIN, fuzzy=False):
+    '''Updates all existing translations (*/LC_MESSAGES/messages.po) based on messages.pot'''
+    generate and translations_generate()
+    command = (['', 'update', '--no-wrap'] + ([] if fuzzy else ['--no-fuzzy-matching']) +
+               [f'--domain={domain}', f'--input-file={TEMPLATE}', f'--output-dir={DIR}'])
+    pybabel().run(command)
+    for locale in new_locales:
+        try:  # trying an update first, to prevent clearing an existing translation
+            pybabel().run(command + ['--init-missing', f'--locale={locale}'])
+        except FileNotFoundError:
+            pybabel().run(['', 'init', '--no-wrap', f'--domain={domain}', f'--input-file={TEMPLATE}',
+                           f'--output-dir={DIR}', f'--locale={locale}'])
+    # handling obsolete entries (removing blank and restoring re-added keys)
+    for locale in os.listdir(DIR):
+        filename = os.path.join(DIR, locale, 'LC_MESSAGES', f'{domain}.po')
+        if os.path.isfile(filename) and os.access(filename, os.W_OK):
+            with open(filename, encoding='utf-8') as fin:
+                text = fin.read()
+            if text != (stripped := replace_obsolete(text)):
+                print(f'processing obsolete entries from catalog {filename}')
+                with open(filename, 'w', encoding='utf-8') as fout:
+                    fout.write(stripped)
+
+def translations_compile(update=False, generate=True, domain=DOMAIN, new_locales=[], fuzzy=False):
+    '''Compiles all existing translations'''
+    update and translations_update(generate=generate, domain=DOMAIN, new_locales=new_locales, fuzzy=fuzzy)
```

Comment:
With correct parameters, a single `translations_compile()` call is sufficient to run all functionality of the script at once